### PR TITLE
EIP-2565 ModExp Gas Cost

### LINKED ISSUE #1972: EIP-2565 for Berlin
Implement https://eips.ethereum.org/EIPS/eip-2565 for Berlin

### PR DESCRIPTION
**Pull request by kclowes**
### What was wrong?

py-evm needs to be updated to support the Berlin Hard fork. This PR implements EIP 2565 which updates the ModExp Gas Cost.

Fixes #1972 

### How was it fixed?
Added a new method to calculate the modexp precompile and added new test vectors accordingly. 

~I also only added some of the test vectors, but am happy to add them all if we want!~ All test vectors have been added!

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [x] Add all test vectors
- [x] Cleanup commits
- [ ] 
[//]: # (For important changes that should go into the release notes please add a newsfragment file as explained here: https://github.com/ethereum/py-evm/blob/master/newsfragments/README.md)

[//]: # (See: https://py-evm.readthedocs.io/en/latest/contributing.html#pull-requests)
- [x] Add entry to the [release notes](https://github.com/ethereum/py-evm/blob/master/newsfragments/README.md)

#### Cute Animal Picture

image

### CODE REVIEW

**Comment by kclowes** [comment]
This is ready for a first review @carver!

**Comment by carver** [comment]
Rebasing on master will get the fix for #1979 to make the istanbul tests pass again

**Comment by carver** [comment]
> I also only added some of the test vectors, but am happy to add them all if we want!

Before digging in, the answer is almost certainly yes :)

**Comment by carver** [comment]
This will fix #1972

**Comment by carver** [comment]
Check out how this precompile cost change was done: https://github.com/ethereum/py-evm/commit/2747d2a4762a6dd271bb9e94fbdd30fd5a8cbba8

I think that keeps the namespace nice and clean, and the fork-specific changes isolated to their respective forks. Can you take a stab at that approach?

**Comment by kclowes** [comment]
> I think that keeps the namespace nice and clean, and the fork-specific changes isolated to their respective forks. Can you take a stab at that approach?

Aha! Yep, that's much cleaner. Will change!

**Comment by kclowes** [comment]
@carver this is ready for another look when you get a chance!

File: eth/constants.py
Comment (carver) [inline eth/constants.py:172]: This can go into `eth/vm/forks/berlin/constants.py` (see the Istanbul fork, for example)

File: eth/precompiles/modexp.py
Comment (carver) [inline eth/precompiles/modexp.py:72]: Don't think this one needs it.
```suggestion
def _compute_modexp_gas_fee_eip_198(data: bytes) -> int:
```

File: eth/precompiles/modexp.py
Comment (carver) [inline eth/precompiles/modexp.py:131]: Maybe `gas_cost` -> `gas_calculator` or something to make it sound a little more method-y? Also, I think the type can be more tightly bound here.

```suggestion
           gas_calculator: Callable[[bytes], int] = _compute_modexp_gas_fee_eip_198) -> BaseComputation:
```

File: eth/precompiles/modexp.py
Comment (carver) [inline eth/precompiles/modexp.py:137]: *nit* I liked these on separate lines, but if you disagree, no need to block the PR over it. :)

File: eth/vm/forks/berlin/computation.py
Comment (carver) [inline eth/vm/forks/berlin/computation.py:39]: Instead of currying here, you could just pass in `_compute_modexp_gas_fee_eip_2565` as the `gas_calculator` method
```suggestion
def _compute_modexp_gas_fee_eip_2565(data: bytes) -> int:
```

File: eth/vm/forks/berlin/computation.py
Comment (carver) [inline eth/vm/forks/berlin/computation.py:59]:
```suggestion
    {force_bytes_to_address(b'\x05'): modexp(gas_calculator=_compute_modexp_gas_fee_eip_2565)},
```

**Review by carver — APPROVED** [review]
Looks good! Just some nitpicks and then ready to merge. :+1:

**Comment by kclowes** [comment]
@carver I just went to merge and realized I can't 😆  This is ready to go when you get a chance, or I'm happy to do it if you give me write permissions!

**Comment by carver** [comment]
I was just wondering that. I'll see if I have permissions to add you.

**Comment by carver** [comment]
Ok, you should have merge access now.

**Pull request merged**
